### PR TITLE
Support multi-output models with bioimageio spec 0.5

### PR DESCRIPTION
**Pull request by alanocallaghan**
Add support for multi-output models using the bioimageio 0.5.x model spec.

We currently handle logits and embeddings by adding them to the measurements of output objects. We don't have a good way to name measurements yet; hopefully we can get axis names at some point in the future of the bioimageio spec. Detection classes are set as PathClass on the output objects.

Detection classes are supported with a [NominalOrOrdinalDataDescr](https://bioimage-io.github.io/spec-bioimage-io/bioimageio_schema_latest/#oneOf_i2_oneOf_i1_inputs_items_data_anyOf_i0). If logits are present too, we try to match the names of outputs and logits. Otherwise, we default to "Class 1", "Logit 1" etc.

It's not totally clear how to handle multiple classes. "Class 1: Class A"?

Semantic segmentation is recognised as an output type, but support is not currently implemented (suggest this could be a future PR for now?).

Superset of #100

The expected model spec is something like the following:

```
outputs:
- axes:
  - {type: batch}
  - channel_names: [nuclei]
    type: channel
  - id: y
    scale: 0.5
    size: {axis_id: y, tensor_id: raw}
    type: space
    unit: micrometer
  - id: x
    scale: 0.5
    size: {axis_id: x, tensor_id: raw}
    type: space
    unit: micrometer
  id: instance_segmentation
  test_tensor: {sha256: 0ab6185a43ccd31b0bfa33f4f5b12a3cc4c0dab07ed3bd521b78bdfd9a8f389d,
    source: test-output_instance_segmentation.npy}
- axes:
  - description: Number of detected objects
    id: n_objects
    size: {}
    type: index
  - {description: Logits for classes, id: logits, size: 19, type: index}
  data:
    type: float32
  id: detection_logits
  test_tensor: {sha256: 0c9a180b08e27a9e5d5bc43d28937747ca526c79c25c4ff647fe7c92a2fa0c95,
    source: test-output_detection_classes.npy}
- axes:
  - description: Number of detected objects
    id: n_objects
    size: {}
    type: index
  - {description: One-hot class labels, id: classes, size: 19, type: index}
  data:
    type: uint8
    values: [fancy_class_0, fancy_class_1, fancy_class_2, fancy_class_3, fancy_class_4,
      fancy_class_5, fancy_class_6, fancy_class_7, fancy_class_8, fancy_class_9, fancy_class_10,
      fancy_class_11, fancy_class_12, fancy_class_13, fancy_class_14, fancy_class_15,
      fancy_class_16, fancy_class_17, fancy_class_18]
  id: detection_classes
  test_tensor: {sha256: 0c9a180b08e27a9e5d5bc43d28937747ca526c79c25c4ff647fe7c92a2fa0c95,
    source: test-output_detection_classes.npy}
```

See https://github.com/qupath/qupath-bioimageio-spec/pull/12

### CODE REVIEW

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java [hunk 1]
```diff
@@ -251,8 +258,64 @@ public Optional<Integer> getNumChannels() {
...
+     */
+    public Optional<List<OutputTensor>> getOutputs() {
+        return getModel().flatMap(model -> Optional.ofNullable(model.getOutputs()));
+    }
+
+    /**
+     * Try to check the output classes from the model spec.
+     * @return The output classes if the model is downloaded, and it's present, otherwise empty.
+     */
+    public List<String> getClasses() {
```
Comment (petebankhead) [inline src/main/java/qupath/ext/instanseg/core/InstanSegModel.java:273]: I think `getClasses()` is too similar to `getClass()`, which has a completely different meaning.

How about `getOutputClasses()`?

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java [hunk 2]
```diff
@@ -251,8 +258,64 @@ public Optional<Integer> getNumChannels() {
...
+    public enum OutputType {
+        // "instance segmentation" "cell embeddings" "cell classes" "cell probabilities" "semantic segmentation"
+        INSTANCE_SEGMENTATION("instance_segmentation"),
+        DETECTION_EMBEDDINGS("detection_embeddings"),
+        DETECTION_LOGITS("detection_logits"),
+        DETECTION_CLASSES("detection_classes"),
+        SEMANTIC_SEGMENTATION("semantic_segmentation");
+
+        private final String type;
+        OutputType(String type) {
```
Comment (petebankhead) [inline src/main/java/qupath/ext/instanseg/core/InstanSegModel.java:297]: Please include a blank line to ensure fields are clearly separated from constructors.
Also make private if it doesn't need to be package private.

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java [hunk 3]
```diff
@@ -251,8 +258,64 @@ public Optional<Integer> getNumChannels() {
...
+                for (var t: elList) {
+                    classes.add(t.toString());
+                }
+            }
+            return classes;
+        }
+        return List.of();
+    }
+
+    public enum OutputType {
```
Comment (petebankhead) [inline src/main/java/qupath/ext/instanseg/core/InstanSegModel.java:288]: Needs a javadoc to explain what it is, and also what each enum means.

(We also use `preferredOutputType` with a different meaning elsewhere)

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java [hunk 4]
```diff
@@ -251,8 +258,64 @@ public Optional<Integer> getNumChannels() {
...
+        private final String type;
+        OutputType(String type) {
+            this.type = type;
+        }
+        @Override
+        public String toString() {
+            return type;
+        }
+
+        public static OutputType fromString(String value) {
```
Comment (petebankhead) [inline src/main/java/qupath/ext/instanseg/core/InstanSegModel.java:305]: Needs javadoc (also explaining that it can return `null`).

File: src/main/java/qupath/ext/instanseg/core/InstanSegOutputToObjectConverter.java [hunk 5]
```diff
@@ -14,21 +16,17 @@
...
 import qupath.lib.roi.interfaces.ROI;
 import qupath.opencv.tools.OpenCVTools;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Random;
+import java.util.*;
```
Comment (petebankhead) [inline src/main/java/qupath/ext/instanseg/core/InstanSegOutputToObjectConverter.java:25]: Trying to convince IntelliJ now to do this...

**Review by petebankhead — COMMENTED** [review]
I haven't tried running it, but added a few comments, mostly related to clarity.

I don't really have an overview of the latest spec and instanseg developments, so I'll go with what you think is best. Will discover if I have any strong opinions after it's merged & I've played around with it some more.

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java (hunk 2 above)
Comment (alanocallaghan) [reply src/main/java/qupath/ext/instanseg/core/InstanSegModel.java:297]: IDE linting tells me private doesn't make any odds here

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java (hunk 4 above)
Comment (alanocallaghan) [reply src/main/java/qupath/ext/instanseg/core/InstanSegModel.java]: Thanks, used optional instead

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java (hunk 3 above)
Comment (alanocallaghan) [reply src/main/java/qupath/ext/instanseg/core/InstanSegModel.java]: Yes, specifyign that it's a tensor type is more clear

File: src/main/java/qupath/ext/instanseg/core/InstanSegModel.java (hunk 1 above)
Comment (alanocallaghan) [reply src/main/java/qupath/ext/instanseg/core/InstanSegModel.java]: Sorry, this is actually unused and should've been removed.

File: src/main/java/qupath/ext/instanseg/core/InstanSegOutputToObjectConverter.java (hunk 5 above)
Comment (alanocallaghan) [reply src/main/java/qupath/ext/instanseg/core/InstanSegOutputToObjectConverter.java]: Yes, I swear I changed the setting the other day...

**Comment by alanocallaghan** [comment]
Going to test this with a few more example models before merging
